### ISSUE TITLE
Implement SendFileAsync feature

### ISSUE THREAD

**Issue by shirhatti** [dotnet/aspnetcore#3807]
For In-Proc

https://github.com/dotnet/aspnetcore/blob/d804e5d6fd6158c3ebc96e1b87788c9b9fc08d36/src/Http/Http.Features/src/IHttpResponseBodyFeature.cs#L44

**Comment by davidfowl** [comment]
I would love to see some basic static files tests automated in the lab before we make this change.

@sebastienros I know we have a benchmark, its just a matter of adding it to the list of benchmarks right?

**Comment by sebastienros** [comment]
We have scenarios. My issue is what we try to measure. What does it mean to have a good static file perf? 

- be able to saturate network at any file size, not cpu
- limit disk access to not saturate on it
- use the least amount of memory
- latency
- compression efficiency

**Comment by davidfowl** [comment]
You're right. I don't know if hammering the server with 1M RPS makes sense for static files. Typical usage might be an initial request for a bunch of resources in parallel. After that they're cached by the client. 

We could grab something like amazon or facebook's client side resources and then send bursts of concurrent connections to access the resources.  Large number of connections, small RPS.

Any thoughts @Tratcher @halter73 @BrennanConroy

**Comment by halter73** [comment]
I think bursts of concurrent connections leading to a large number of connections and small RPS will still quickly become bandwidth constrained after a small initial ramp up. If we still use RPS as the metric for this kind test, it will probably benefit algorithms with less initial batching because smaller initial batches leads to less latency and a faster ramp up to saturating the NICs.

Another metric we should keep in mind in addition to the list @sebastienros listed is fairness. We wouldn't want to switch to an algorithm that used less CPU by working on larger batches at the cost of unfairly slowing down unlucky connections. This is especially true for static file serving scenarios with a lot of connections where it seems unlikely the server will be CPU constrained than memory constrained.
